# Assignment 4 - AM5801
**Name:** Atharv Shete  
**Date:** September 10, 2025

## Question 1

#### Part (a): Reading input and displaying the initial orders

We will:
- Read the number of pending orders `n`
- Read the current `n` revenue values
- Read the threshold value `x`
- Read the new order revenue `r`
- Display the current list of orders


In [24]:
n = int(input("Enter number of pending orders: "))
orders = list(
    map(
        float, input("Enter the revenues of current orders (space-separated): ").split()
    )
)
x = float(input("Enter the threshold value: "))
r = float(input("Enter the new order revenue: "))

# Display the current orders
print("Current pending orders:", orders)
print("Threshold value:", x)
print("New order revenue:", r)


Current pending orders: [10.0, 8.25, 9.75, 5.0]
Threshold value: 7.5
New order revenue: 5.5


#### Part (b): Greedy insertion strategy

Steps:
1. Try inserting the new order `r` at each possible position (from start to end).
2. After inserting, check **all consecutive triples** in the new list.
3. If all triples satisfy the threshold condition, accept the order and stop.
4. If no valid insertion exists, reject the order.

This follows a **greedy approach** because:
- We do not search for the "best" position overall.
- Instead, we take the **first valid position** found and immediately accept it.


In [25]:
def is_valid(order_list, x):
    for i in range(len(order_list) - 2):
        if (order_list[i] + order_list[i + 1] + order_list[i + 2]) / 3 < x:
            return False
    return True


accepted = False
final_orders = []

# Try inserting at each position
for i in range(n + 1):  # n+1 possible positions
    new_list = orders[:i] + [r] + orders[i:]
    if is_valid(new_list, x):
        final_orders = new_list
        accepted = True
        break


#### Part (c): Display result

We will:
- Show the updated list if the order is accepted.
- Otherwise, print "Order rejected".

### Why this is greedy?
The algorithm is greedy because:
- It accepts the new order at the **first valid position** encountered.
- It does not backtrack or try to optimize further.
- Once a valid solution is found, it stops searching.

In [26]:
if accepted:
    print("Updated orders after insertion:", final_orders)
else:
    print("Order rejected")

Updated orders after insertion: [5.5, 10.0, 8.25, 9.75, 5.0]


## Question 2: Fibonacci Sequence - Naive Recursion vs Dynamic Programming

#### Part (a): Naive recursive Fibonacci

We will:
- Implement a recursive Fibonacci function.
- Count the number of function calls made.
- Note: This will grow exponentially with `n`.


In [28]:
naive_calls = 0


def fib_naive(n):
    global naive_calls
    naive_calls += 1
    if n == 0:
        return 0
    elif n == 1:
        return 1
    else:
        return fib_naive(n - 1) + fib_naive(n - 2)


# Example run
n = 10
naive_calls = 0
fib_naive(n)
print(f"Naive recursion calls for F({n}):", naive_calls)


Naive recursion calls for F(10): 177


#### Part (b): Dynamic programming Fibonacci

We will:
- Use bottom-up tabulation to compute Fibonacci.
- Count only the number of **additions** performed.
- Complexity = O(n).


In [29]:
def fib_dp(n):
    if n == 0:
        return 0, 0
    if n == 1:
        return 1, 0

    dp = [0] * (n + 1)
    dp[0], dp[1] = 0, 1
    additions = 0

    for i in range(2, n + 1):
        dp[i] = dp[i - 1] + dp[i - 2]
        additions += 1  # each step is one addition

    return dp[n], additions


# Example run
fib_val, dp_additions = fib_dp(n)
print(f"DP additions for F({n}):", dp_additions)


DP additions for F(10): 9


#### Part (c): Comparison

We will now compare:
- Naive recursion calls
- DP additions

for different values of `n`.


In [32]:
def compare_methods(max_n=15):
    print(f"{'n':<5}{'Naive Calls':<15}{'DP Additions'}")
    print("-" * 35)
    for test_n in range(max_n):
        global naive_calls
        naive_calls = 0
        fib_naive(test_n)  # compute and count calls
        naive_call_count = naive_calls

        _, dp_adds = fib_dp(test_n)
        print(f"{test_n:<5}{naive_call_count:<15}{dp_adds}")


# Example comparison for n = 0..14
compare_methods(15)


n    Naive Calls    DP Additions
-----------------------------------
0    1              0
1    1              0
2    3              1
3    5              2
4    9              3
5    15             4
6    25             5
7    41             6
8    67             7
9    109            8
10   177            9
11   287            10
12   465            11
13   753            12
14   1219           13


#### Part (d): Efficiency Discussion

- **Naive recursion** makes an exponential number of calls:  
  For n = 10 → 177 calls, for n = 20 → 13529 calls, etc.
- **Dynamic programming** only requires (n-1) additions.  
  For n = 10 → 9 additions, for n = 20 → 19 additions.

### Efficiency Improvement:
Dynamic programming reduces the complexity from **O(2^n)** to **O(n)**.  
This is a dramatic efficiency improvement, especially as `n` grows large.


## Question 3

#### Part (a): Reading input and displaying the list


In [41]:
# Read input
n = int(input("Enter number of elements: "))
arr = list(map(int, input("Enter the elements (space-separated): ").split()))

# Display initial list
print("Original list:", arr)


Original list: [4, 3, 2, 1]


#### Part (b): Merge Sort with inversion counting

We extend merge sort:
- Recursively divide the list into halves.
- Count inversions in left half, right half, and during merge.
- While merging, if an element from the right sublist is smaller than an element in the left sublist,
  then **all remaining elements in the left sublist** form inversions with it.

This gives O(n log n) complexity.

In [ ]:

def merge_and_count(left, right):
    """Merge two sorted lists and count inversions."""
    merged = []
    i = j = 0
    inversions = 0
    
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
            inversions += (len(left) - i)  # all remaining elements in left are > right[j]
    
    # Add remaining elements
    merged.extend(left[i:])
    merged.extend(right[j:])
    
    return merged, inversions


def merge_sort_and_count(arr):
    """Return sorted array and inversion count."""
    if len(arr) <= 1:
        return arr, 0
    
    mid = len(arr) // 2
    left, left_inv = merge_sort_and_count(arr[:mid])
    right, right_inv = merge_sort_and_count(arr[mid:])
    merged, split_inv = merge_and_count(left, right)
    
    total_inversions = left_inv + right_inv + split_inv
    return merged, total_inversions

# Example run
sorted_arr, inv_count = merge_sort_and_count(arr)


#### Part (c): Display results and explanation

We will:
- Print the sorted list
- Print the inversion count

##### Why is this divide-and-conquer?
- The problem is **divided** into two halves (like merge sort).
- Inversions are **conquered** by solving in subproblems and combining counts in the merge step.
- This reduces complexity from **O(n²)** to **O(n log n)**, showcasing the power of divide-and-conquer.


In [40]:
print("Sorted list:", sorted_arr)
print("Total number of inversions:", inv_count)


Sorted list: [1, 2, 3, 4, 5]
Total number of inversions: 3
